blog/_posts/2019-07-14-multithreading.md

```diff
@@ -0,0 +1,464 @@
+
+## Some history
+
+One of the most surprising aspects of this new feature is just how long it has been in
+the works.
+From the very beginning --- prior even to the 0.1 release --- Julia has had the `Task`
+type, providing symmetric coroutines and event-based I/O.
+So we have always had a unit of *concurrency* in the language, it just wasn't *parallel*
+(simultaneous streams of execution) yet.
+We knew we needed parallelism though, so in 2014 (roughly the version 0.3 timeframe) we
+set about the long process of making all of our code thread-safe.
+Yichao Yu put in some particularly impressive work on the garbage collector and signal
+handling.
+Kiran Pamnany (of Intel) put some basic infrastructure in place for starting and
+running multiple threads.
+
+Within about two years, we were ready to release the `@threads` macro in version 0.5,
+which provides simple parallel loops.
+Even though that wasn't the final design we wanted, it did two important jobs:
+it let Julia programmers start taking advantage of multiple cores, and provided
+test cases to shake out thread-related bugs in our runtime.
+`@threads` had some huge limitations, however.
+`@threads` loops could not be nested: all the functions you call from within such a loop
+must not themselves use `@threads`.
+It was also incompatible with our `Task` and I/O system: you couldn't do any I/O or
+switch among `Task`s inside a threaded loop.
+
+So the next logical step was to merge the `Task` and threading systems, and "simply"
+(cue laughter) allow `Task`s to run simultaneously on a pool of threads.
+We had many early discussions with Arch Robison (then also of Intel) and concluded
+that this was the best model for our language.
+After version 0.5 (around 2016) Kiran started experimenting with a new parallel
+task scheduler [PARTR][] based on the idea of depth-first scheduling.
+He sold all of us on it with some nice animated slides, and it also didn't hurt that
+he was willing to do some of the work.
+The plan was to first develop PARTR as a standalone C library so it could be tested
+and benchmarked on its own, and then integrate it with the Julia runtime.
+
+After Kiran completed the standalone version of PARTR, we embarked on a series of
+work sessions including Anton Malakhov (also of Intel) to figure out how to do
+the integration.
+The Julia runtime brings many extra features, such as garbage collection and
+event-based I/O, so this was not entirely straightforward.
+Somewhat disappointingly, though not unusually for a complex software project,
+it took much longer than expected --- nearly two years --- to get the new
+system working reliably.
+A later section of this post will explain some of the internals and difficulties
+involved for the curious.
+But first, let's take it for a spin.
+
+## How to use it
+
+To use Julia with multiple threads, set the `JULIA_NUM_THREADS` environment
+variable:
+
+```
+$ JULIA_NUM_THREADS=4 ./julia
+```
```

Comment:
Note that IDEs like Juno automatically detect the number of cores in a user's processor, giving Julia multithreading out of the box when used in these systems.

Reply:
Even if you automatically detect cores, you still need to be able to change the number - to give some cores to BLAS, or the OS or something else. So, ideally, Juno etc. would present the cores available, but have a way for you to change it in the IDE.

Reply:
It does.

![Capture](https://user-images.githubusercontent.com/1814174/61190941-48d65780-a672-11e9-8bd6-25cfc2470eea.PNG)